File: src/main/java/com/github/dkoval/leetcode/challenge/ReverseBits.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/reverse-bits/">Reverse Bits</a>
 * <p>
 * Reverse bits of a given 32 bits signed integer.
 * <p>
 * Constraints:
 * <ul>
 *  <li>0 <= n <= 2^31 - 2</li>
 *  <li>n is even.</li>
 * </ul>
 */
public interface ReverseBits {

    int reverseBits(int n);

    class ReverseBitsRev1 implements ReverseBits {

        @Override
        public int reverseBits(int n) {
            var x = n;
            var res = 0;
            var i = 31;
            while (x > 0) {
                final var bit = x & 1;
                res |= bit << i;
                x >>= 1;
                i--;
            }
            return res;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/ReverseBits.kt
package com.github.dkoval.leetcode.challenge

object ReverseBitsBitByBit: ReverseBits {

    override fun reverseBits(n: Int): Int {
        var result = 0
        var bits = n
        var pos = 31
        while (bits != 0) {
            val bit = bits and 1
            result = result or (bit shl pos)
            bits = bits ushr 1
            pos--
        }
        return result
    }
}

object ReverseBitsMaskAndShift: ReverseBits {

    override fun reverseBits(n: Int): Int {
        var result = (n ushr 16) or (n shl 16)
        result = ((result and 0xff00ff00.toInt()) ushr 8) or ((result and 0x00ff00ff) shl 8)
        result = ((result and 0xf0f0f0f0.toInt()) ushr 4) or ((result and 0x0f0f0f0f) shl 4)
        result = ((result and 0xcccccccc.toInt()) ushr 2) or ((result and 0x33333333) shl 2)
        result = ((result and 0xaaaaaaaa.toInt()) ushr 1) or ((result and 0x55555555) shl 1)
        return result
    }
}
